Handled rate limiting in public CI/CD tool

drewsonne opened the pull request:
I had an issue where I was experiencing rate limiting when trying to install glide on my travis build instance (https://travis-ci.org/drewsonne/go-gocd/jobs/289153125#L649) so I've added an option which will check for environment variables `$GITHUB_USER` and `$GITHUB_TOKEN` and if it finds them, will use them as basic for auth wget and curl.

drewsonne committed:
Added github auth to address rate limiting

drewsonne committed:
Added auth to get

mattfarina commented:
I like the idea of working around the rate limiting. I wonder if the GraphQL API (v4) would provide a way to work around that for everyone... https://developer.github.com/v4/

I'm just thinking out loud

drewsonne commented:
The general idea is that when you call the github API without auth tokens in travis (or any other free OS build service), you're sharing the same IP/CIDR anonymous "user" as all the other OS builds, s hitting the rate limit is super easy.

I'm having a read through the rate limit in v4, now.

drewsonne commented:
Actually... I think the problem is it's hitting the `api.github.com` endpoint which is rate limited. 

Say, for example, the below release archive.

`https://github.com/Masterminds/glide/releases/download/v0.13.0/glide-v0.13.0-linux-arm64.tar.gz`

`glide.sh` has the tag name, `GOARCH`, `GOOS`, so as far as I can see (please correct me if I'm wrong), there's no specific need to hit that API endpoint.

At the moment, you check if there's a distribution archive, and *then* download the file. An alternative would be to attempt the file download and if you get a 4xx, then do the distribution check, and provide error info to the end user.

In theory, this would make one less call on successful downloads as well :-)

mattfarina commented:
GraphQL does rate limiting differently. See https://developer.github.com/v4/guides/resource-limitations/

franciscocpg commented:
About github rating limit, see this:

https://github.com/Masterminds/glide/issues/639#issuecomment-255095785
> @mattfarina, On the glide side, we are making the call to the github API just to get the latest download URL. Maybe you should publish this URL to the https://github.com/Masterminds/glide.sh repository (like you do with version for example) so glide get script doesn't have to deal with this github API rate limits.

https://github.com/Masterminds/glide/issues/639#issuecomment-255183683
> I think @sdboyer is right about the API limit. they've removed the limit for some requests but it's likely better to just put the data on glide.sh as @franciscocpg suggested.

IMO the solution mentioned in the thread above is still valid. And should be simple to implement :smiley:

drewsonne commented:
@mattfarina Yup, the rate limiting is now on the complexity of the query, rather than a simple 1 "rate unit" query. But the problem is that I am sharing my rate quota with everyone else on travis-ci/circle-ci/etc. So even if I improve the rate limit "count" for glide, I can't control the rate limiting for everyone.

I will have a go at hitting the release URL directly and then checking for the latest. That being said... if in the error case (ie, release for a GOOS/GOARCh k/v doesn't exist) I still have to hit the api endpoint... I'm still gonna be rate limited.

mattfarina commented on get:
```diff
@@ -71,24 +71,34 @@ initDownloadTool() {
 }
 
+initGithubAuth() {
+	if [ ! -z "$GITHUB_USER$GITHUB_TOKEN" ]; then
```
What if `$GITHUB_USER` has a value and `$GITHUB_TOKEN` is empty? When using hub there are cases where someone could supply `$GITHUB_PASSWORD` instead.

These two values should be checked separately to make sure both are not empty.

drewsonne replied:
The GitHub token is used in the same way as the password, so there's no need for a separate env var. You would use the password in place of the token. 

That said, I'm going to look at changing the PR as per the discussion above so we don't hit the API endpoint except in the event of an error, so those vars won't be needed to install. If an error occurs, the rate-limiting may still be an issue, but the problem becomes "we can't provide a more informative error because we're rate limited" instead of "we can't install because we're rate limited". In that case, the GitHub may be required and in that event, your suggestion would be reasonable to split the check out.

drewsonne committed:
Revert "Added auth to get"

This reverts commit 5f2eee4cf7e2c8acb2d36318583286a1bc892028.

drewsonne committed:
Revert "Added github auth to address rate limiting"

This reverts commit c61a73bd442d854565a8052d7b9ec7c284f95d2a.

drewsonne committed:
Added non-api endpoint for download

drewsonne committed:
Modified release download

Hit the non-rate-limited release endpoint, rather than the API as the first attempt. If the file could not be found, query the API and ask for the latest release

drewsonne committed:
Fixed inconsistent whitespace

drewsonne committed:
Moved logging to relevant location

drewsonne committed:
Removed duplicate logging

drewsonne commented:
I have made changes as we discussed above, please let me know your thoughts on this

mattfarina commented:
@drewsonne Thanks for the contribution

drewsonne commented:
thank you! :-)
